README.md
# WorkdayHelper
Count workdays and list holiday events in a range of dates with PHP taking care of public holidays and other custom closing days.

Inspired by Massimo Simonini getWorkdays() Gist.
@see     https://gist.github.com/massiws/9593008

Benefits and features:
- you can specify a range of working days in a week (from Monday to Sunday);
- you can add custom holidays and business closures, for example as a result of a database query;
- it returns a calendar of holidays for that specific range of dates;
- you can use your custom recursive holiday calendar (see `$publicHolidays`);
- it takes care of the timezone of your application;
- it automatically calculates Easter, and Easter mondays taking care of the timezone (requires PHP with `ext-calendar`);

## Requirements

PHP 5.6.0 and later

## Composer

The preferred way to install this extension is through [composer](http://getcomposer.org/download/).

Either run

```composer require --prefer-dist letsjump/workday-helper-php```


or add


```"letsjump/workday-helper-php": "~1.0.0"```

to the `require` section of your composer.json.

## Manual Installation

If you do not wish to use Composer, you can download the [latest release](https://github.com/letsjump/workday-helper-php).

```php
require_once('/path/to/workday-helper-php/WorkdayHelper.php');
```

## Dependencies

### PHP ext-calendar

To automatically calculate Easter dates, you have to compile PHP with `--enable-calendar`. See [ext-calendar](https://www.php.net/manual/en/book.calendar.php). You can refer to [this Stack Overflow question](https://stackoverflow.com/questions/5297894/fatal-error-call-to-undefined-function-easter-date/51609625) if you are in a Docker environment.

If your PHP isn't compiled with ext-calendar and if you cannot compile it, please set `$calculateEaster` to `false` otherwhise WorkingdayHelper will throw an exception.

 ## Usage:

1. Count the days worked in January while working from Monday to Friday, taking care of public holidays:
```php
use letsjump\workdayHelper\WorkdayHelper;

$closingDays                 = new WorkdayHelper('2021-01-01', '2021-01-31');
$closingDays->workingDays    = [1, 2, 3, 4, 5];

echo $closingDays->getWorkdays(); // (19 days)
```

2. count the day worked in april while working Monday, Wednesday and Friday, taking care of public holidays:

```php
$closingDays                 = new WorkdayHelper('2021-04-01', '2021-04-30');
$closingDays->workingDays    = [1, 3, 5];
echo $closingDays->getWorkdays(); // (12 days)
```

3. Add a strike to the custom closing days

```php
$closingDays                 = new WorkdayHelper('2021-01-01', '2021-01-31');
$closingDays->workingDays          = [1, 2, 3, 4, 5];
$closingDays->customClosing = [
     [
         'date'    => '2021-01-18',
         'event'   => 'Strike!',
         'options' => [
             'employee_id'        => 345,
             'htmlClass' => 'deep-purple'
         ]
     ],
];
echo $closingDays->getWorkdays(); // (18 days)
```

4. Get the calendar with all the closing days for a specific date interval

```php
$closingDays                 = new WorkdayHelper('2021-01-01', '2021-12-31');
$closingDays->workingDays          = [0, 1, 2, 3, 4, 5, 6]; // don't forget to set every day of the week!

<table>
<?php foreach ($closingDays->getCalendar() as $holiday): ?>
<tr>
     <td><?= $holiday['date'] ?></td>
     <td><?= $holiday['event'] ?></td>
</tr>
<?php endforeach ?>
</table>
```

![Gii module](docs/assets/images/calendar.jpg)

### Add custom closings

Custom closure is an array of events. Each event is an array with this configuration:

key | value | mandatory
----|-------|----------
date| date in Y-m-d format | yes
event | a string representing the name of the closing | yes
options | a string or an array passed as is to the holiday calendar. It may contain information such as event_id, user_id, html attributes, etc.. Please remember that data are passed as they are so, to prevent malicious injection attacks, consider the use of a string purifier function | no

__example__

```php

$myInstantatedClass->customClosing = [
         [
             'date'  => '2021-01-10',
             'event' => 'Chiusura per ferie'
         ],
         [
             'date'    => '2021-01-05',
             'event'   => 'Strike!',
             'options' => [
                  'id'        => 345,
                  'htmlClass' => 'green'
             ]
         ],
          ...
];
```

### Replace the default recursive holiday calendar

The default recursive holiday calendar is an array of events. Each event is an array with this configuration:

key | value | mandatory
----|-------|----------
m-d | the date must be in m-d format | yes
event | a string representing the name of the holiday | yes
options | a string or an array passed as is to the holiday calendar. It may contain information such as event_id, user_id, html attributes, etc.. Please remember that data are passed as they are so, to prevent malicious injection attacks, consider the use of a string purifier function | no

__example__

```php

$myInstantatedClass->publicHolidays = [
         [
             'm-d'  => '12-25',
             'event' => 'Christmas'
         ],
         [
             'm-d'    => '12-26',
             'event'   => 'Boxing day',
             'options' => [
                  'id'        => 345,
                  'htmlClass' => 'green'
             ]
         ],
          ...
];
```

### Holiday calendar
The holiday calendar returns a list of holidays for the working days in the given date range.
So if you want to retrive _all the closing days_ in that date range you should set all the days of the week into the $workingDays Array E.G. $myWorkDay->workingDays[0,1,2,3,4,5,6].

The output is an array of events. Each array has the unix timestamp of the holiday as key and contains:

key | value
----|------
unixTimestamp | the unix timestamp of the event
date | the readable date of the event (you can format it with the parameter `$outputFormat`
event | the name of the event type | `public` or `custom` depending if it is a publicHoliday or a customClosing.
options | a string or an array of user custom options

Example

```php
      [
         [1609455600] => [
             [unixTimestamp] => 1609455600,
             [date] => 2021-01-01, # control the format with $outputFormat property
             [event] => Capodanno, # description of the event
             [type] => public, #public / custom
             [options] => # custom option passed by the $customClosing or the $publicHolidays Array
         ],
         ...
      ]
```

## Side notes

The recursive holiday and the default date format are localized for Italy. I know it's not the right way to work, but I need to quickly develop an application for a customer. If you want to use your localization, please refer to *Replace the default recursive holiday calendar*. If you plan to use intensively in your application, you may extend the base class and configure there the basic properties.

Please report any bug or whishes into the apposite [issue tracker](https://github.com/letsjump/workday-helper-php/issues)

changelog.md
### CHANGE LOG

**v1.0.0**
- initial release

src/WorkdayHelper.php
<?php

namespace letsjump\workdayHelper;

use yii\base\InvalidConfigException;

/**
 * Class WorkdayHelper
 *
 * Count work days and list holiday events in a range of dates with PHP taking care of public holidays and other custom
 * closing days.
 * Inspired by Massimo Simonini getWorkdays() Function. See https://gist.github.com/massiws/9593008
 *
 * @author  Gianpaolo Scrigna <letsjump@gmail.com>
 */
class WorkdayHelper
{
    public const TYPE_PUBLIC = 'public';
    public const TYPE_CUSTOM = 'custom';
    
    /**
     * @var int[] days to consider as worked
     * in a week, where sunday == 0 and saturday == 6
     * @see
     */
    public $workingDays = [1, 2, 3, 4, 5];
    
    /**
     * @var string date format for the closing days output list
     */
    public $outputFormat = 'Y-m-d';
    
    /**
     * @var bool calculate and add the easter dates
     *           to the closing days output list
     */
    public $calculateEaster = true;
    
    /**
     * @var array[] of custom closures.
     *            Any custom closure array need at least the keys:
     *            - date ([date] a date in the Y-m-d format)
     *            - event ([string] name of the event)
     * The optional array key `options` can contain any custom variable you need
     * and it will be passed as is to the closing days output list
     */
    public $customClosing = [];
    
    /**
     * @var array[] array of public holiday dates where:
     *               key: [date] date in m-d format
     *               value: [string] name of the event
     *
     */
    public $publicHolidays = [
        [
            'm-d'     => '01-01',
            'event'   => 'Capodanno',
        ],
        [
            'm-d'   => '01-06',
            'event' => 'Epifania'
        ],
        [
            'm-d'   => '04-25',
            'event' => 'Festa della Liberazione'
        ],
        [
            'm-d'   => '05-01',
            'event' => 'Festa del Lavoro'
        ],
        [
            'm-d'   => '06-02',
            'event' => 'Festa della Repubblica'
        ],
        [
            'm-d'   => '08-15',
            'event' => 'Ferragosto'
        ],
        [
            'm-d'   => '11-01',
            'event' => 'Ognissanti'
        ],
        [
            'm-d'   => '12-08',
            'event' => 'Immacolata'
        ],
        [
            'm-d'   => '12-25',
            'event' => 'Natale'
        ],
        [
            'm-d'   => '12-26',
            'event' => 'Santo Stefano'
        ],
    ];
    
    private $startDateObject;
    private $endDateObject;
    private $years = [];
    private $closing = [];
    private $workdays = null;
    private $holidays = [];
    
    public function __construct($startDate, $endDate)
    {
        try {
            $this->startDateObject = new \DateTime($startDate);
            $this->endDateObject   = new \DateTime($endDate);
        } catch (\Exception $e) {
            var_dump($e->getMessage());
            exit;
        }
        $this->getYearsInterval();
        
    }
    
    /**
     * @return integer the number of the worked days between the interval of dates.
     */
    public function getWorkdays()
    {
        if ($this->workdays === null) {
            $this->run();
        }
        
        return $this->workdays;
    }
    
    /**
     * @return array the array of all the closing days between the interval of dates.
     * @throws \InvalidArgumentException
     */
    public function getCalendar()
    {
        if ($this->workdays === null) {
            $this->run();
        }
        
        return $this->holidays;
    }
    
    /**
     * Fill the array $this->years with every year from the date interval passed
     */
    private function getYearsInterval()
    {
        for ($year = $this->startDateObject->format('Y'); $year <= $this->endDateObject->format('Y'); $year++) {
            $this->years[] = $year;
        }
    }
    
    /**
     * @param \DateTime $dateObject
     * @param string $description
     * @param string $type
     * @param null $options
     *
     * Add an item to $this->closing array
     */
    private function addClosing($dateObject, $description, $type, $options = null)
    {
        $unixTimestamp                 = $dateObject->format('U');
        $this->closing[$unixTimestamp] = [
            'unixTimestamp' => $unixTimestamp,
            'date'          => $dateObject->format($this->outputFormat),
            'event'         => $description,
            'type'          => $type,
            'options'       => $options
        ];
    }
    
    /**
     * Add the public holidays to the closing Array
     *
     * @throws \InvalidArgumentException
     */
    private function addPublicHolidays()
    {
        foreach ($this->years as $year) {
            foreach ($this->publicHolidays as $holiday) {
                try {
                    if ( ! array_key_exists('m-d', $holiday) || ! array_key_exists('event', $holiday)) {
                        throw new \InvalidArgumentException('Malformed PublicHoliday array. m-d or event key doesn\'t exists');
                    }
                    $dateObject = new \DateTime($year . '-' . $holiday['m-d']);
                    $options    = isset($holiday['options']) ? $holiday['options'] : null;
                    $this->addClosing($dateObject, $holiday['event'], self::TYPE_PUBLIC, $options);
                } catch (\Exception $e) {
                    var_dump($e->getMessage());
                }
            }
            if ($this->calculateEaster === true) {
                $this->addEasterDates($year);
            }
        }
    }
    
    /**
     * @param integer $year
     *
     * @throws \InvalidArgumentException
     *
     * Calculate the easter days for the year passed
     */
    private function addEasterDates($year)
    {
        try {
            if(function_exists('easter_days')) {
                $equinox      = new \DateTime($year . "-03-21");
                $easterObject = $equinox->add(new \DateInterval('P' . easter_days($year) . 'D'));
                $this->addClosing($easterObject, 'Pasqua', self::TYPE_PUBLIC);
                $easterMondayObject = $easterObject->add(new \DateInterval('P1D'));
                $this->addClosing($easterMondayObject, 'Lunedì dell\'Angelo', self::TYPE_PUBLIC);
            } else {
                throw new InvalidConfigException("ext-calendar not found in your PHP installation");
            }
        } catch (\Exception $e) {
            var_dump($e->getMessage());
        }
    }
    
    /**
     * Add the custom closing day to the closing Array
     */
    private function addCustomClosing()
    {
        if ( ! empty($this->customClosing)) {
            foreach ($this->customClosing as $closure) {
                try {
                    if ( ! array_key_exists('date', $closure) || ! array_key_exists('event', $closure)) {
                        throw new \InvalidArgumentException('Malformed CustomClosure array. Date or event key doesn\'t exists');
                    }
                    if (($dateObject = new \DateTime($closure['date'])) !== false) {
                        $options = isset($closure['options']) ? $closure['options'] : null;
                        $this->addClosing($dateObject, $closure['event'], self::TYPE_CUSTOM, $options);
                    }
                } catch (\Exception $e) {
                    var_dump($e->getMessage());
                }
            }
        }
    }
    
    /**
     * Calculate the closing days, the number of days worked and the closing days calendar.
     *
     * @throws \InvalidArgumentException
     */
    private function run()
    {
        $this->addPublicHolidays();
        $this->addCustomClosing();
        $this->workdays = 0;
        for (
            $unixDay = $this->startDateObject->format('U'); $unixDay <= $this->endDateObject->format('U'); $unixDay = strtotime("+1 day",
            $unixDay)
        ) {
            $dayOfWeek = date("w", $unixDay);
            if (in_array((int)$dayOfWeek, $this->workingDays, true)) {
                if ( ! array_key_exists($unixDay, $this->closing)) {
                    $this->workdays++;
                } else {
                    $this->holidays[$unixDay] = $this->closing[$unixDay];
                }
            }
        }
    }
}